packages/reactivity/src/ast-analyzer.ts

```diff
@@ -0,0 +1,122 @@
+import * as path from 'node:path'
+import type { DeepnoteBlock } from '@deepnote/blocks'
+import { z } from 'zod'
+import { safelyCallChildProcessWithInputOutput } from './child-process-utils'
+
+export class AstAnalyzerInternalError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'AstAnalyzerInternalError'
+  }
+}
+
+const AST_ANALYZER_SUPPORTED_CELL_TYPES_LIST = [
+  'code',
+  'sql',
+  'button',
+  'big-number',
+  'notebook-function',
+  'input-text',
+  'input-textarea',
+  'input-file',
+  'input-select',
+  'input-date',
+  'input-date-range',
+  'input-slider',
+  'input-checkbox',
+  'input-number',
+  'input-dropdown',
+] as const
+
+export interface BlockContentDepsWithOrder {
+  id: string
+  definedVariables: string[]
+  usedVariables: string[]
+  importedModules?: string[]
+  error?: {
+    type: string
+    message: string
+  }
+  order: number
+}
+
+const SUPPORTED_CELL_TYPES = new Set<string>(AST_ANALYZER_SUPPORTED_CELL_TYPES_LIST)
+
+// Zod schema for the AST analyzer Python script output
+const AstAnalyzerItemSchema = z.object({
+  id: z.string(),
+  definedVariables: z.array(z.string()),
+  usedVariables: z.array(z.string()),
+  importedModules: z.array(z.string()).optional(),
+  error: z
+    .object({
+      type: z.string(),
+      message: z.string(),
+    })
+    .optional(),
+})
+
+const AstAnalyzerSuccessSchema = z.array(AstAnalyzerItemSchema)
+const AstAnalyzerErrorSchema = z.object({ errorMessage: z.string() })
+const AstAnalyzerResponseSchema = z.union([AstAnalyzerSuccessSchema, AstAnalyzerErrorSchema])
+
+/**
+ * Tested within dag.test.ts integration test.
+ * Intentionally mixing Python and SQL blocks here because we want to use just one local process invocation.
+ */
+export async function getBlocksContentDeps(blocks: DeepnoteBlock[]): Promise<BlockContentDepsWithOrder[]> {
+  const blocksNeedingComputation = blocks
+    // Process only supported cell types
+    .filter(block => SUPPORTED_CELL_TYPES.has(block.type))
+    // Format blocks for parser
+    .map(block => ({
+      ...block,
+      content: block.content ?? '',
+    }))
+
+  // Early return if no blocks need computation
+  if (blocksNeedingComputation.length === 0) {
+    return []
+  }
+
+  try {
+    const inputData = JSON.stringify({ blocks: blocksNeedingComputation })
+
+    const scriptPath = path.join(__dirname, './scripts/ast-analyzer.py')
```

Comment:
Use `join` to construct the entire path

```suggestion
    const scriptPath = path.join(__dirname, 'scripts', 'ast-analyzer.py')
```